Photon and ValueMap

nsmith- opened the pull request:
I took the seed cluster from the photon and passed it through the tool, and numbers came out!
I'll let my ntupler run for a bit to see how they look, but also here is a ValueMap implementation that could be used for MINIAOD when you integrate this into cmssw.  I let the electron implementation to you ;)

nsmith- committed:
Allow photons and make a ValueMap producer

Perhaps ElectronIDHelper could be called EgammaIDHelper or something.
Also not sure if the photon seed cluster is necessarily the best choice.

nsmith- committed:
More variables in photon valuemaps

Also correct some documentation

artlbv commented:
Hi Nick, great that it works! And thanks for the value map — we’check that for the electrons as well.
However, I don’t think it’s a good idea to use the same ID helper for electrons and photons as now: the track variables are very different and mixing the accessors does not seem to be safe to me. 
What do you think?

nsmith- commented:
Well yeah, storing an electron pointer and then not checking null is asking for trouble.
As I mentioned in a commit message, perhaps ElectronIDHelper could be called EgammaIDHelper or something and the accessors can have a `nullptr` guard around them in case it is not an electron in the current iteration.

jkiesele commented:
I agree with Nick, calling it EgammaIDHelper makes more sense in this case.
But for now, I would just add the guards. I can do a refactoring PR after that if you want.

artlbv commented on EgammaAnalysis/interface/LongDeps.h:
```diff
@@ -14,6 +14,7 @@ class LongDeps{
         inline float energyFH() const {return energyFH_;}
         inline float energyBH() const {return energyBH_;}
         inline const std::vector<float>& energyPerLayer() const {return energyPerLayer_;}
+        inline float energy(size_t layer) const {return (layer<layers_.size()) ? energyPerLayer_[layer] : 0.;}
```
Why this extra member? `energyPerLayer_` is initialized with 0s for all layers, and the layers.size check is incorrect: e.g. the cluster can contain layer 30 while spanning only 20 layers.

nsmith- replied:
The function signature in VARIABLES.md was something like this, so I assumed it was just missing. By all means please fix this as necessary, I was just guessing.

nsmith- replied:
I guess it should be `energyPerLayer_.size()`

nsmith- commented on EgammaAnalysis/plugins/HGCalPhotonIDValueMapProducer.cc:
```diff
@@ -0,0 +1,189 @@
+      maps_["nLayers"].push_back(ld.nLayers());
+      maps_["firstLayer"].push_back(ld.firstLayer());
+      maps_["lastLayer"].push_back(ld.lastLayer());
+      maps_["firstLayerEnergy"].push_back(ld.energy(ld.firstLayer()));
```
So is this correct?

nsmith- committed:
Fix range check

jkiesele commented:
Hi all,
is there anything else open in this PR?

jkiesele commented:
Dear all,
after some research, I unfortunately just discovered that the map as implemented here will not work as it is for the miniAOD production. But I found an example of how it should be implemented and it should not require too many changes:
https://insight.io/github.com/cms-sw/cmssw/blob/master/RecoEgamma/PhotonIdentification/plugins/PhotonIDValueMapProducer.cc

In particular this line will be helpful:
https://insight.io/github.com/cms-sw/cmssw/blob/master/RecoEgamma/PhotonIdentification/plugins/PhotonIDValueMapProducer.cc?line=443

nsmith- committed:
Use edm::View<reco::Photon>

So it can also work with pat photons (e.g. MINIAOD production)

beaudett commented on EgammaAnalysis/plugins/HGCalPhotonIDValueMapProducer.cc:
```diff
@@ -67,6 +67,17 @@ HGCalPhotonIDValueMapProducer::HGCalPhotonIDValueMapProducer(const edm::Paramete
+  maps_["nLayers"] = {};
+  maps_["firstLayer"] = {};
+  maps_["lastLayer"] = {};
+  maps_["firstLayerEnergy"] = {};
```
Hello, 
sorry for the late reaction. I am not sure that "firstLayerEnergy" is a very good variable for ID. Layer per layer energies can fluctuate (especially in the first layer for photons), and I suspect it is quite sensitive to PU. 
F.

artlbv committed:
Duplicate eleID helper for photons

artlbv committed:
Add photon helper as for electrons

artlbv committed:
Remove PhotonFWD to make compile

artlbv committed:
Add test for photons

artlbv committed:
Test on photon samples

nsmith- committed:
Add Jan's isolation code

nsmith- committed:
Merge branch 'artlbv' into photon

Merging PR 4 with the split of photon and electron helper

nsmith- committed:
Remove photon stuff from ElectronIDHelper

nsmith- committed:
Add Jan's isolation ring sums to value map

Also remove firstLayerEnergy per Florian's comment
Also kill all evil carriage returns in Jan's files

nsmith- commented on EgammaAnalysis/src/PhotonHGCalIsoProducer.cc:
```diff
@@ -0,0 +1,108 @@
+        }
+    }
+
+    //does this make sense here or should it be done later?
```
@jkiesele I think leaving the raw energy might be safer than the ratio.  For example, the run 2 photon isolation cuts are only weakly pt-dependent.  Anyway the photon energy stored in that variable could be at some point a regression energy or some other thing that is not just sum(rechits).

jkiesele replied:
Fully agree with you.

nsmith- committed:
Don't divide isolation deposits by photon energy

nsmith- committed:
Remove -O0 optimization

artlbv commented:
Ok, looks good to me. Will try to add @beaudett's electron value map soon as well.
